### PR TITLE
[7X] Remove legacy gppkg.

### PR DESCRIPTION
**Pull request by higuoxing**
This patch contains 3 seperate commits:

1. Some extensions under `contrib/` seems shipped using the legacy gppkg. The first commit removes the legacy gppkg support from these extensions. (a2be4b4f8d7784d66d2459636974d2a164c7f0fb)
2. The second commit removes the legacy gppkg itself from repo. (3f003b616a3277d49c6073c58c87f6635e17806e)
3. The third commit removes legacy gppkg tests from pipelines. I haven't regenerated the `gpdb_main-generated.yml`, I will do that after this PR being reviewed. (Test pipeline https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/remove-legacy-gppkg-from-gpdb-rocky8) (90a9ba58a026cbbfbd6d1d17ca9a435413cbeab9)
4. Re-generate prod pipeline (07553452c5ba37b070451364037d323ff29087c1)

Notes:
1. Some utilities relies on the function of syncing packages of the legacy gppkg. E.g., `gpexpand`, `gprecoversegment`, etc. Packages will be synced according to their metadata.
2. Legacy gppkg's documentation is **not** removed in this PR.

### CODE REVIEW

File: gpMgmt/bin/gppylib/operations/test/unit/__init__.py [hunk 1]
```diff
@@ -1,13 +0,0 @@
-# Make sure Python loads the modules of this package via absolute paths.
```
Comment (beeender) [inline gpMgmt/bin/gppylib/operations/test/unit/__init__.py:1]: seems to be needed still

File: gpMgmt/bin/gppylib/operations/test/unit/test_unit_buildMirrorSegments.py [hunk 2]
```diff
@@ -1,1122 +0,0 @@
-#!/usr/bin/env python3
```
Comment (beeender) [inline gpMgmt/bin/gppylib/operations/test/unit/test_unit_buildMirrorSegments.py:1]: ditto

File: gpMgmt/bin/gppylib/operations/test/unit/test_unit_configurationImplGpdb.py [hunk 3]
```diff
@@ -1,94 +0,0 @@
-#!/usr/bin/env python3
```
Comment (beeender) [inline gpMgmt/bin/gppylib/operations/test/unit/test_unit_configurationImplGpdb.py:1]: ditto

File: gpMgmt/bin/gppylib/operations/test/unit/test_unit_deletesystem.py [hunk 4]
```diff
@@ -1,23 +0,0 @@
-#!/usr/bin/env python3
```
Comment (beeender) [inline gpMgmt/bin/gppylib/operations/test/unit/test_unit_deletesystem.py:1]: ditto

File: gpMgmt/bin/gppylib/operations/test/unit/test_unit_initstandby.py [hunk 5]
```diff
@@ -1,60 +0,0 @@
-#!/usr/bin/env python3
```
Comment (beeender) [inline gpMgmt/bin/gppylib/operations/test/unit/test_unit_initstandby.py:1]: ditto

File: gpMgmt/bin/gppylib/operations/test/unit/test_unit_ping.py [hunk 6]
```diff
@@ -1,52 +0,0 @@
-from mock import patch
```
Comment (beeender) [inline gpMgmt/bin/gppylib/operations/test/unit/test_unit_ping.py:1]: ditto

File: gpMgmt/bin/gppylib/operations/test/unit/test_unit_reload.py [hunk 7]
```diff
@@ -1,247 +0,0 @@
-#!/usr/bin/env python3
```
Comment (beeender) [inline gpMgmt/bin/gppylib/operations/test/unit/test_unit_reload.py:1]: ditto

File: gpMgmt/bin/gppylib/operations/test/unit/test_unit_segment_reconfigurer.py [hunk 8]
```diff
@@ -1,89 +0,0 @@
-import datetime
```
Comment (beeender) [inline gpMgmt/bin/gppylib/operations/test/unit/test_unit_segment_reconfigurer.py:1]: ditto

File: gpMgmt/bin/gppylib/operations/test/unit/test_unit_segment_tablespace_locations.py [hunk 9]
```diff
@@ -1,36 +0,0 @@
-#!/usr/bin/env python3
```
Comment (beeender) [inline gpMgmt/bin/gppylib/operations/test/unit/test_unit_segment_tablespace_locations.py:1]: ditto

File: gpMgmt/bin/gppylib/operations/test/unit/test_unit_update_pg_hba_on_segments.py [hunk 10]
```diff
@@ -1,177 +0,0 @@
-#!/usr/bin/env python3
```
Comment (beeender) [inline gpMgmt/bin/gppylib/operations/test/unit/test_unit_update_pg_hba_on_segments.py:1]: ditto

File: gpMgmt/bin/gppylib/operations/test/unit/test_unit_utils.py [hunk 11]
```diff
@@ -1,100 +0,0 @@
-#!/usr/bin/env python3
```
Comment (beeender) [inline gpMgmt/bin/gppylib/operations/test/unit/test_unit_utils.py:1]: ditto

**Review by beeender — APPROVED** [review]

File: gpAux/Makefile [hunk 12]
```diff
@@ -502,7 +502,6 @@ SET_VERSION_SCRIPTS = \
 	lib/python/gppylib/programs/clsHostCacheLookup.py \
```
Comment (liuxueyang) [inline gpAux/Makefile:502]: Do we need to remove the `gppkg` word in `gpAux/BUILD_INSTRUCTIONS`?

Comment (higuoxing) [reply gpAux/Makefile:502]: Good catch! I think we should update the instruction.

**Review by liuxueyang — APPROVED** [review]
LGTM

File: gpMgmt/test/behave/mgmt_utils/environment.py [hunk 13]
```diff
@@ -70,11 +70,6 @@ def before_feature(context, feature):
         dbconn.execSQL(context.conn, 'insert into t3 values(1, 4)')
         context.conn.commit()
 
-    if 'gppkg' in feature.tags:
-        run_command(context, 'bash demo/gppkg/generate_sample_gppkg.sh buildGppkg')
-        run_command(context, 'cp -f /tmp/sample-gppkg/sample.gppkg test/behave/mgmt_utils/steps/data/')
```
Comment (liuxueyang) [inline gpMgmt/test/behave/mgmt_utils/environment.py:75]: Should the file `gpMgmt/test/behave/mgmt_utils/steps/data/sample.gppkg` be removed?

**Review by liuxueyang — APPROVED** [review]

File: gpMgmt/bin/gppylib/operations/package.py
[the remark this answers is not in the document]
Comment (higuoxing) [reply gpMgmt/bin/gppylib/operations/package.py:1]: I added this file back to keep `SyncPackages` competible with what we have before.

File: gpAux/Makefile (hunk 12 above)
Comment (higuoxing) [reply gpAux/Makefile:502]: I'd like to do it in a follow-up PR. Since there is some documentation mentions gppkg as well.

**Comment by higuoxing** [comment]
I've tested it manually. If the package's metadata is correct, package files can be synced to segments successfully. If the metadata file is broken, it will print the error stack and tell user use `gppkg --clean` after the expansion finishing.

```
20230428:07:11:07:024686 gpexpand:cdw:gpadmin-[INFO]:-The following packages will be installed on sdw3: a
20230428:07:11:07:024686 gpexpand:cdw:gpadmin-[ERROR]:-Unexpect version of package's metadata of gppkg: 'a'
Traceback (most recent call last):
  File "/usr/local/greenplum-db-devel/lib/python/gppylib/commands/base.py", line 277, in run
    self.cmd.run()
  File "/usr/local/greenplum-db-devel/lib/python/gppylib/operations/__init__.py", line 53, in run
    self.ret = self.execute()
  File "/usr/local/greenplum-db-devel/lib/python/gppylib/operations/package.py", line 96, in execute
    file_list = get_package_file_list(package)
  File "/usr/local/greenplum-db-devel/lib/python/gppylib/operations/package.py", line 55, in get_package_file_list
    raise Exception("Unexpect version of package's metadata of gppkg: '{}'".format(gppkg_package_name))
Exception: Unexpect version of package's metadata of gppkg: 'a'
20230428:07:11:07:024686 gpexpand:cdw:gpadmin-[ERROR]:-Syncing of Greenplum Database extensions has failed.
Traceback (most recent call last):
  File "/usr/local/greenplum-db-devel/bin/gpexpand", line 1951, in sync_packages
    operation.get_ret()
  File "/usr/local/greenplum-db-devel/lib/python/gppylib/operations/__init__.py", line 65, in get_ret
    raise self.ret
  File "/usr/local/greenplum-db-devel/lib/python/gppylib/commands/base.py", line 277, in run
    self.cmd.run()
  File "/usr/local/greenplum-db-devel/lib/python/gppylib/operations/__init__.py", line 53, in run
    self.ret = self.execute()
  File "/usr/local/greenplum-db-devel/lib/python/gppylib/operations/package.py", line 96, in execute
    file_list = get_package_file_list(package)
  File "/usr/local/greenplum-db-devel/lib/python/gppylib/operations/package.py", line 55, in get_package_file_list
    raise Exception("Unexpect version of package's metadata of gppkg: '{}'".format(gppkg_package_name))
Exception: Unexpect version of package's metadata of gppkg: 'a'
20230428:07:11:07:024686 gpexpand:cdw:gpadmin-[WARNING]:-Please run gppkg --clean after successful expansion.
20230428:07:11:07:024686 gpexpand:cdw:gpadmin-[INFO]:-Locking catalog
20230428:07:11:07:024686 gpexpand:cdw:gpadmin-[INFO]:-Locked catalog
```
.

**Review by beeender — APPROVED** [review]

File: gpMgmt/bin/gppylib/operations/package.py [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+            logger.info('The packages on {} are consistent.'.format(self.host))
+            return
+
+        if install_package_set:
+            logger.info(
+                'The following packages will be installed on {}: {}'.format(self.host, ', '.join(sorted(install_package_set))))
+            for package in install_package_set:
+                logger.debug('Copying {} to {}'.format(package, self.host))
+                file_list = self.get_package_file_list(package)
+                sync_list_file = os.path.join(GPHOME, "{}.synclist".format(package))
```
Comment (Sasasu) [inline gpMgmt/bin/gppylib/operations/package.py:97]: put this file to `/$TMPDIR/gpdb_<package>_synclist_<random char * 4>` would be better.

`Operation` or `RemoteOperation` can emit an exception, then the `RemoveFile()` part will not be executed.
then a secret unknown usage file will be left in GPHOME.

File: gpMgmt/bin/gppylib/operations/package.py [hunk 15]
```diff
@@ -46,6 +46,19 @@ def dereference_symlink(path):
 GPPKG_METADATA_EXTENSION = ".ini"
 GPPKG_METADATA_FILELINE_PATTERM = re.compile(r""".*permission.*type.*size.*sha1.*md5""")
 
+def get_package_file_list(gppkg_package_name):
+    filelist = []
+    gppkg_metadata_file = "{}/{}{}".format(GPPKG_METADATA_DIR, gppkg_package_name, GPPKG_METADATA_EXTENSION)
+    with open(gppkg_metadata_file, "rb") as f:
+        for linenum, line in enumerate(f):
+            if linenum == 0 and line != b'version=v1\n':
```
Comment (Sasasu) [inline gpMgmt/bin/gppylib/operations/package.py:56]:
```suggestion
            if linenum == 0 and not line.startswith(b'version=v1'):
```

this file may be a binary file, the only promise is this file starts with `version=v1`. (don't check the new line

and please note, if the file is a binary file. read by line will read the entire file into memory.  if the file is a bad file, the program will OOM.
the common way is to read small data first, then start decode.

```
f = open("", "rb");
d = f.read(10);
if !d.startwith(b"version=v1"):
    emit error
f.close()
f = open("", "rt")
```

but, the current code is sufficient and does not need to be changed

**Review by Sasasu — APPROVED** [review]

File: gpMgmt/bin/gppylib/operations/package.py (hunk 15 above)
Comment (higuoxing) [reply gpMgmt/bin/gppylib/operations/package.py:56]: THERE'S ONLY TWO THINGS I HATE IN THIS WORLD.

PEOPLE WHO DON'T CARE USERS' INVALID INPUTS AND ... PEOPLE WHO PUT A BINARY FILE INTO `$GPHOME/usr/local/gppkg`  WHICH ENDS WITH `.ini`

**Pull request merged**